perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java

```diff
@@ -4137,6 +4145,44 @@ private static boolean objectAndRoleManageableByEntity(String entityToManage, Pe
 		return false;
 	}
 
+	/**
+	 * Checks if setting role requires MFA based on configuration in perun-roles, throws MfaPrivilegeException if requirements unmet.
+	 * @param assigningEntity user or group to which the role will be assigned (User)
+	 * @param complementaryObject which will be bounded with the role (Vo)
+	 * @param role role name (VOADMIN)
+	 * @throws MfaPrivilegeException if MFA requirements are not met
+	 */
+	private static void checkMfaForSettingRole(PerunSession sess, PerunBean assigningEntity, PerunBean complementaryObject, String role) {
+		RoleManagementRules rules;
+		try {
+			rules = AuthzResolverImpl.getRoleManagementRules(role);
+		} catch (RoleManagementRulesNotExistsException e) {
+			throw new InternalErrorException("Management rules not exist for the role " + role, e);
+		}
+
+		List<Map<String, String>> mfaPolicies = rules.getAssignmentCheck();
+		Map<String, Set<Integer>> mapOfBeans = new HashMap<>();
+
+		if (mfaPolicies != null && !mfaPolicies.isEmpty() && complementaryObject != null) {
+			mapOfBeans = fetchAllRelatedObjects(Arrays.asList(complementaryObject));
+		}
+
+		// check complementary objects for MFA requirements
+		if (!mfaAuthorized(sess, mfaPolicies, mapOfBeans)) {
+			throw new MfaPrivilegeException("Multi-Factor authentication is required - bounded object is critical.");
```

Comment:
Maybe providing the complementary object in question in this exception could prove useful (we can call `setRole()` with a list of multiple complementary objects and then we wouldn't know because of which object the exception was thrown).